velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,223 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "PrefixSort.h"
+
+namespace facebook::velox::exec {
+
+namespace {
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize() {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize<TypeKind::BIGINT>() {
+  return sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+}
+
+FOLLY_ALWAYS_INLINE uint32_t normalizeSize(const TypeKind& typeKind) {
+  return VELOX_DYNAMIC_TYPE_DISPATCH(normalizeSize, typeKind);
+}
+
+template <TypeKind typeKind>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  VELOX_UNSUPPORTED("PrefixSort not support the type.");
+}
+
+template <>
+FOLLY_ALWAYS_INLINE void normalizeRowColumn<TypeKind::BIGINT>(
+    const PrefixSortLayout& prefixSortLayout,
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* FOLLY_NONNULL const row,
+    char* FOLLY_NONNULL const prefix) {
+  using T = TypeTraits<TypeKind::BIGINT>::NativeType;
+  // Store null as min/max value according compare flags.
+  if (RowContainer::isNullAt(row, rowColumn.nullByte(), rowColumn.nullMask())) {
+    CompareFlags compareFlags = prefixSortLayout.compareFlag(index);
+    PrefixSortEncoder::encode(
+        ((compareFlags.ascending && compareFlags.nullsFirst) ||
+         (!compareFlags.ascending && !compareFlags.nullsFirst))
+            ? std::numeric_limits<T>::min()
+            : std::numeric_limits<T>::max(),
+        prefix + prefixSortLayout.prefixOffset(index));
+  } else {
+    PrefixSortEncoder::encode(
+        *(reinterpret_cast<T*>(row + rowColumn.offset())),
+        prefix + prefixSortLayout.prefixOffset(index));
+  }
+  // Invert bits if desc
+  if (!prefixSortLayout.compareFlag(index).ascending) {
+    for (uint64_t s = 0; s < sizeof(T); s++) {
+      *(prefix + prefixSortLayout.prefixOffset(index) + s) =
+          ~*(prefix + prefixSortLayout.prefixOffset(index) + s);
+    }
+  }
+}
+} // namespace
+
+PrefixSortLayout::PrefixSortLayout(
+    const std::vector<TypePtr>& types,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const uint32_t maxPrefixKeyLength)
+    : normalizedKeySize_(0),
+      numNormalizedKeys_(0),
+      numSortKeys_(types.size()),
+      keyCompareFlags_(keyCompareFlags) {
+  VELOX_CHECK_GE(types.size(), 0);
+  for (TypePtr type : types) {
+    if (numNormalizedKeys_ > maxPrefixKeyLength) {
```

Comment:
I expect maxPrefixKeyLength to specify maximum number of bytes to use for a combined normalized byte buffer. Here, it looks like maxPrefixKeyLength is maximum number of sorting keys to include in the normalized byte buffer. Is this intentional?

Reply:
>  Is this intentional?

No , it is a mistake.